File: api/schemas/update_domain/update_domain.py
import graphene

from graphql import GraphQLError

from app import logger
from db import db_session
from functions.auth_wrappers import require_token
from functions.auth_functions import is_user_write
from functions.input_validators import cleanse_input, cleanse_input_list
from functions.slugify import slugify_value
from models import Domains
from scalars.url import URL
from scalars.selectors import Selectors


class UpdateDomainInput(graphene.InputObjectType):
    """
    Input object used to define the argument fields for the updateDomain
    mutation
    """

    current_url = URL(
        description="The current domain that is being requested to be updated.",
        required=True,
    )
    updated_url = URL(
        description="The new domain you wish to update the current domain to be.",
        required=True,
    )
    updated_selectors = Selectors(
        description="The new DKIM selector strings corresponding to this domain",
        required=False,
    )


class UpdateDomain(graphene.Mutation):
    """
    Mutation allows the modification of domains if domain is updated through
    out its life-cycle
    """

    class Arguments:
        input = UpdateDomainInput(
            required=True,
            description="updateDomain input object containing all arguement fields.",
        )

    status = graphene.Boolean()

    @require_token
    def mutate(self, info, **kwargs):
        user_id = kwargs.get("user_id")
        user_roles = kwargs.get("user_roles")
        current_domain = cleanse_input(kwargs.get("input", {}).get("current_url"))
        updated_domain = cleanse_input(kwargs.get("input", {}).get("updated_url"))
        updated_selectors = cleanse_input_list(
            kwargs.get("input", {}).get("updated_selectors", [])
        )

        # Check to see if current domain exists
        domain_orm = Domains.query.filter(Domains.domain == current_domain).first()

        if domain_orm is None:
            logger.warning(
                f"User: {user_id} tried to update {current_domain} but the domain does not exist."
            )
            raise GraphQLError("Error, unable to update domain.")

        if is_user_write(user_roles=user_roles, org_id=domain_orm.organization_id):
            Domains.query.filter(Domains.domain == current_domain).update(
                {
                    "domain": updated_domain,
                    "selectors": updated_selectors,
                    "slug": slugify_value(updated_domain),
                }
            )

            try:
                db_session.commit()
                logger.info(f"User: {user_id} successfully updated {domain_orm.domain}")
                return UpdateDomain(status=True)
            except Exception as e:
                db_session.rollback()
                db_session.flush()
                logger.error(
                    f"A database error occurred when user: {user_id} tried to update {current_domain} error: {str(e)}"
                )
                return UpdateDomain(status=False)
        else:
            logger.warning(
                f"User: {user_id} tried to update {current_domain} but does not have permissions to {domain_orm.organization.slug}."
            )
            raise GraphQLError("Error, unable to update domain.")

File: api-js/src/mutations/domain/update-domain.js
const { GraphQLID, GraphQLNonNull, GraphQLList } = require('graphql')
const { mutationWithClientMutationId, fromGlobalId } = require('graphql-relay')
const { t } = require('@lingui/macro')
const { Domain, Selectors } = require('../../scalars')
const { domainType } = require('../../types')

const updateDomain = new mutationWithClientMutationId({
  name: 'UpdateDomain',
  description:
    'Mutation allows the modification of domains if domain is updated through out its life-cycle',
  inputFields: () => ({
    domainId: {
      type: GraphQLNonNull(GraphQLID),
      description: 'The global id of the domain that is being updated.',
    },
    orgId: {
      type: GraphQLNonNull(GraphQLID),
      description:
        'The global ID of the organization used for permission checks.',
    },
    domain: {
      type: Domain,
      description: 'The new url of the of the old domain.',
    },
    selectors: {
      type: new GraphQLList(Selectors),
      description:
        'The updated DKIM selector strings corresponding to this domain.',
    },
  }),
  outputFields: () => ({
    domain: {
      type: domainType,
      description: 'The updated domain.',
      resolve: async (payload) => {
        return payload.domain
      },
    },
  }),
  mutateAndGetPayload: async (
    args,
    {
      i18n,
      query,
      collections,
      transaction,
      userKey,
      auth: { checkPermission, userRequired },
      validators: { cleanseInput },
      loaders: { domainLoaderByKey, orgLoaderByKey },
    },
  ) => {
    const { id: domainId } = fromGlobalId(cleanseInput(args.domainId))
    const { id: orgId } = fromGlobalId(cleanseInput(args.orgId))
    const updatedDomain = cleanseInput(args.domain)

    let selectors
    if (typeof args.selectors !== 'undefined') {
      selectors = args.selectors.map((selector) => cleanseInput(selector))
    } else {
      selectors = null
    }

    // Get User
    await userRequired()

    // Check to see if domain exists
    const domain = await domainLoaderByKey.load(domainId)

    if (typeof domain === 'undefined') {
      console.warn(
        `User: ${userKey} attempted to update domain: ${domainId}, however there is no domain associated with that id.`,
      )
      throw new Error(i18n._(t`Unable to update domain. Please try again.`))
    }

    // Check to see if org exists
    const org = await orgLoaderByKey.load(orgId)

    if (typeof org === 'undefined') {
      console.warn(
        `User: ${userKey} attempted to update domain: ${domainId} for org: ${orgId}, however there is no org associated with that id.`,
      )
      throw new Error(i18n._(t`Unable to update domain. Please try again.`))
    }

    // Check permission
    const permission = await checkPermission({ orgId: org._id })

    if (
      permission !== 'user' &&
      permission !== 'admin' &&
      permission !== 'super_admin'
    ) {
      console.warn(
        `User: ${userKey} attempted to update domain: ${domainId} for org: ${orgId}, however they do not have permission in that org.`,
      )
      throw new Error(i18n._(t`Unable to update domain. Please try again.`))
    }

    // Check to see if org has a claim to this domain
    let countCursor
    try {
      countCursor = await query`
        FOR v, e IN 1..1 ANY ${domain._id} claims 
          FILTER e._from == ${org._id}
          RETURN e
      `
    } catch (err) {
      console.error(
        `Database error occurred while user: ${userKey} attempted to update domain: ${domainId}, error: ${err}`,
      )
      throw new Error(i18n._(t`Unable to update domain. Please try again.`))
    }

    if (countCursor.count < 1) {
      console.warn(
        `User: ${userKey} attempted to update domain: ${domainId} for org: ${orgId}, however that org has no claims to that domain.`,
      )
      throw new Error(i18n._(t`Unable to update domain. Please try again.`))
    }

    // Generate list of collections names
    const collectionStrings = []
    for (const property in collections) {
      collectionStrings.push(property.toString())
    }

    // Setup Transaction
    const trx = await transaction(collectionStrings)

    // Update domain
    const domainToInsert = {
      domain: updatedDomain || domain.domain,
      lastRan: domain.lastRan,
      selectors: selectors || domain.selectors,
    }

    try {
      await trx.run(
        async () =>
          await query`
          UPSERT { _key: ${domain._key} }
            INSERT ${domainToInsert}
            UPDATE ${domainToInsert}
            IN domains
      `,
      )
    } catch (err) {
      console.error(
        `Transaction run error occurred when user: ${userKey} attempted to update domain: ${domainId}, error: ${err}`,
      )
      throw new Error(i18n._(t`Unable to update domain. Please try again.`))
    }

    // Commit transaction
    try {
      await trx.commit()
    } catch (err) {
      console.error(
        `Transaction commit error occurred when user: ${userKey} attempted to update domain: ${domainId}, error: ${err}`,
      )
      throw new Error(i18n._(t`Unable to update domain. Please try again.`))
    }

    // Clear dataloader and load updated domain
    await domainLoaderByKey.clear(domain._key)
    const returnDomain = await domainLoaderByKey.load(domain._key)

    console.info(`User: ${userKey} successfully updated domain: ${domainId}.`)
    returnDomain.id = returnDomain._key

    return {
      domain: {
        ...returnDomain,
      },
    }
  },
})

module.exports = {
  updateDomain,
}
